### PR TITLE
Shader_Ir: Miscellaneous features for future implementations

### PR DESCRIPTION
**Pull request by FernandoS27**
This provides a set of features needed for further implementations into Shader Tracking and code deriving such as array samplers. This adds two small features: Determining information from the in use gpu driver and saving bound buffer used to the shader disk cache.

This PR, sadly, invalidates previous shader caches as this changes are needed for future fixes.

### CODE REVIEW

File: src/video_core/guest_driver.cpp [hunk 1]
```diff
@@ -0,0 +1,36 @@
+// Copyright 2019 yuzu Emulator Project
```
Comment (bunnei) [inline src/video_core/guest_driver.cpp]: 2020?

File: src/video_core/guest_driver.cpp [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+#include "video_core/guest_driver.h"
+
+namespace VideoCore {
+
+void GuestDriverProfile::DeduceTextureHandlerSize(std::vector<u32>&& bound_offsets) {
+    if (texture_handler_size_deduced) {
+        return;
+    }
+    std::size_t size = bound_offsets.size();
```
Comment (bunnei) [inline src/video_core/guest_driver.cpp]: Can be `const`

File: src/video_core/guest_driver.cpp [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    if (size < 2) {
+        return;
+    }
+    std::sort(bound_offsets.begin(), bound_offsets.end(),
+              [](const u32& a, const u32& b) { return a < b; });
+    u32 min_val = 0xFFFFFFFF; // set to highest possible 32 bit integer;
```
Comment (bunnei) [inline src/video_core/guest_driver.cpp]: <climits>  ?

File: src/video_core/guest_driver.h [hunk 4]
```diff
@@ -0,0 +1,39 @@
+// Copyright 2019 yuzu Emulator Project
```
Comment (bunnei) [inline src/video_core/guest_driver.h]: 2020?

File: src/video_core/guest_driver.h [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+
+#include "common/common_types.h"
+
+namespace VideoCore {
+
+/**
+ * The GuestDriverProfile class is used to learn about the GPU drivers behavior and collect
+ * information necessary for impossible to avoid HLE methods like shader tracks.
+ */
+class GuestDriverProfile {
```
Comment (bunnei) [inline src/video_core/guest_driver.h]: Shouldn't this be HostDriveProfile? (i.e. the users PC is the "host", the emulated hardware (switch) is the "Guest")

File: src/video_core/shader/const_buffer_locker.h [hunk 6]
```diff
@@ -49,6 +52,10 @@ class ConstBufferLocker {
     /// Inserts a bindless sampler key.
     void InsertBindlessSampler(u32 buffer, u32 offset, Tegra::Engines::SamplerDescriptor sampler);
 
+    /// Set the bound buffer for this locker.
+
```
Comment (bunnei) [inline src/video_core/shader/const_buffer_locker.h]: Remove extra line

File: src/video_core/shader/decode.cpp [hunk 7]
```diff
@@ -315,4 +315,25 @@ u32 ShaderIR::DecodeInstr(NodeBlock& bb, u32 pc) {
     return pc + 1;
 }
 
+void ShaderIR::PostDecode() {
+    // Deduce texture handler size if needed
+    auto* gpu_driver = locker.AccessGuestDriverProfile();
```
Comment (bunnei) [inline src/video_core/shader/decode.cpp]: if (!gpu_driver) {
    // Log error?
    return;
}

(early-return code is easier to read than many nested if's)

File: src/video_core/guest_driver.cpp (hunk 3 above)
Comment (bunnei) [reply src/video_core/guest_driver.cpp]: * use climits

File: src/video_core/shader/decode.cpp (hunk 7 above)
Comment (FernandoS27) [reply src/video_core/shader/decode.cpp]: yeah but on design I don't want early return as different unrelated passes may be run on this function

File: src/video_core/guest_driver.h (hunk 5 above)
Comment (FernandoS27) [reply src/video_core/guest_driver.h]: This gets info on the driver used by the game not the one in the PC. Therefore, it's a guest driver.

File: src/video_core/guest_driver.cpp [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+void GuestDriverProfile::DeduceTextureHandlerSize(std::vector<u32>&& bound_offsets) {
+    if (texture_handler_size_deduced) {
+        return;
+    }
+    const std::size_t size = bound_offsets.size();
+    if (size < 2) {
+        return;
+    }
+    std::sort(bound_offsets.begin(), bound_offsets.end(),
+              [](const u32& a, const u32& b) { return a < b; });
```
Comment (ReinUsesLisp) [inline src/video_core/guest_driver.cpp]: Use `std::less{}` instead of a lambda.

File: src/video_core/rasterizer_interface.h [hunk 9]
```diff
@@ -78,5 +79,12 @@ class RasterizerInterface {
     /// Initialize disk cached resources for the game being emulated
     virtual void LoadDiskResources(const std::atomic_bool& stop_loading = false,
                                    const DiskResourceLoadCallback& callback = {}) {}
+
+    GuestDriverProfile& AccessGuestDriverProfile() {
```
Comment (ReinUsesLisp) [inline src/video_core/rasterizer_interface.h]: Add a documentation commentary here for consistency.

File: src/video_core/shader/shader_ir.h [hunk 10]
```diff
@@ -187,6 +187,7 @@ class ShaderIR final {
     };
 
     void Decode();
+    void PostDecode();
```
Comment (ReinUsesLisp) [inline src/video_core/shader/shader_ir.h]: Marking it for removal, read above.

File: src/video_core/guest_driver.h [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+
+#include <vector>
+
+#include "common/common_types.h"
+
+namespace VideoCore {
+
+/**
+ * The GuestDriverProfile class is used to learn about the GPU drivers behavior and collect
+ * information necessary for impossible to avoid HLE methods like shader tracks.
```
[the remark this answers is not in the document]
Comment (FernandoS27) [reply src/video_core/guest_driver.h]: You can't solve entscheidung problems.

File: src/video_core/shader/shader_ir.cpp [hunk 12]
```diff
@@ -27,6 +27,7 @@ ShaderIR::ShaderIR(const ProgramCode& program_code, u32 main_offset, CompilerSet
                    ConstBufferLocker& locker)
     : program_code{program_code}, main_offset{main_offset}, settings{settings}, locker{locker} {
     Decode();
+    PostDecode();
```
[the remark this answers is not in the document]
Comment (FernandoS27) [reply src/video_core/shader/shader_ir.cpp]: No, it makes no logic to for the methods in PostDecode to be within Decode as normally they have nothing to do with Decoding a Shader but more to do with learning from it.

**Pull request closed without merging**